### ISSUE TITLE
cmake default install prefix is wrong

### ISSUE THREAD

**Issue by bjornfor** [NixOS/nixpkgs#24215]
## Issue description
According to cmake documentation, the default installation prefix on UNIX should be ```/usr/local```, but cmake from nixpkgs uses ```/var/empty/local/```.

**Comment by bjornfor** [comment]
We should use sane defaults.

**Comment by vcunat** [comment]
I'm not sure we should encourage installing in other ways than through nix.  It seems e.g. likely that the projects will pick up dependencies on some /nix/store paths and those can be later garbage-collected.

**Comment by bjornfor** [comment]
I don't think we're "encouraging" anything by fixing that path. I use this for testing the installation procedure of a project while working in qtcreator. I don't plan to use this for "permanent" installs.

I don't think we should make tools more difficult to use as a way to "help" users. If we want maximum "help" we might as well inject startup code in all dev tools that abort execution unless there is an indication in the environment that it is being run under nix-build or nix-shell...

**Comment by vcunat** [comment]
I'd expect users to choose something under `$HOME` instead of a global path, but that's just my personal opinion...

**Comment by bjornfor** [comment]
Right, that's what I did too (so I didn't have to use 'sudo').

But that isn't a reason for deviating from upstream default value and "breaking" their documentation.

**Comment by Mic92** [comment]
yes, this might be confusing, if somebody uses cmake to build a package for a non-nixos system on nixos for instance.

**Comment by bjornfor** [comment]
Or just using cmake from nixpkgs on non-NixOS.

**Comment by drever** [comment]
This issue is a bit confusing. I was trying to install my own project on macOS with the cmake installation from nix and it took me some time to find this thread. Does nix use a special build of cmake which has the changed default path?

The makefile generated by cmake offers the possibility to [set the destination directory](https://cmake.org/cmake/help/v3.0/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX) like this:

     make DESTDIR=myDestDir install

Using this doesn't really work since the `/var/empty` is appended after the cmake install prefix, i.e it results in `myDestDir/var/empty/local/lib`. Copying the files manually to `/local/lib/` works as a workaround.

**Comment by bjornfor** [comment]
I'm guessing the /var/empty thing comes from a combination of [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/cmake/setup-hook.sh#L10) and [this](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/build-managers/cmake/default.nix#L64).

**Comment by bjornfor** [comment]
Still important to me.

**Comment by Mic92** [comment]
should not be solved.

**Comment by Mic92** [comment]
I think usually you want to avoid mixing cmake files from legacy package managers with nix on non-nixos systems. Which is why I think this replacement was put into place. For the same reason programs are not reading /etc/ld.so.conf because they might pick up stuff that they were not meant to read.

**Comment by spraharsh** [comment]
Related. It might help to alert the user since this behavior was unexpected while trying to debug an issue elsewhere http://sundials.2283335.n4.nabble.com/Memory-error-while-running-cvode-petsc-examples-tt4655872.html

**Comment by veprbl** [comment]
Just hit this. I think it has to do with what is discussed in #118781

**Comment by smasher164** [comment]
I noticed that my source of breakage here was that my cmake file was checking if the CMAKE_INSTALL_PREFIX was set to the default before overriding it. However, since nix sets its own CMAKE_INSTALL_PREFIX, this branch never got triggered. I ended up just removing the `if`, but I wonder if there's a better way to deal with this.
```
if(CMAKE_INSTALL_PREFIX_INITIALIZED_TO_DEFAULT)
  set(CMAKE_INSTALL_PREFIX "${BUILD_BUNDLE_DIR}" CACHE PATH "..." FORCE)
endif()
```

**Comment by pithuene** [comment]
Thanks @smasher164!
Just hit this issue while creating a new Flutter project and this if condition is present in the default linux build configuration.
Removing it fixed my issue :+1:
